keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/ScalarDoubleTensor.java

```diff
@@ -157,21 +159,17 @@ public DoubleTensor times(double that) {
 
     @Override
     public DoubleTensor matrixMultiply(DoubleTensor that) {
-        if (that.isLengthOne()) {
+        if (shape.length == 2) {
             return that.times(value);
+        } else {
+            throw new IllegalArgumentException("Cannot matrix multiply shape: " + Arrays.toString(shape));
         }
-        throw new IllegalArgumentException("Cannot use matrix multiply with scalar. Use times instead.");
     }
 
     @Override
     public DoubleTensor tensorMultiply(DoubleTensor that, int[] dimsLeft, int[] dimsRight) {
-        if (that.isLengthOne()) {
-            if (dimsLeft.length > 1 || dimsRight.length > 1 || dimsLeft[0] != 0 || dimsRight[0] != 0) {
-                throw new IllegalArgumentException("Tensor multiply sum dimensions out of bounds for scalar");
-            }
-            return that.times(value);
-        }
-        throw new IllegalArgumentException("Cannot use tensor multiply with scalar. Use times instead.");
+        long[] resultShape = getTensorMultiplyResultShape(this.shape, that.getShape(), dimsLeft, dimsRight);
+        return Nd4jDoubleTensor.create(value, shape).tensorMultiply(that, dimsLeft, dimsRight).reshape(resultShape);
```

Comment:
Good point. I was falling back to using the tensor multiply in Nd4j to avoid having to calculate the result shape. But I ended up doing that anyway because Nd4j doesn't calculate it correctly for one of my test cases :(